core/new-gui/src/app/workspace/component/workflow-editor/workflow-editor.component.spec.ts

```diff
@@ -859,5 +872,69 @@ describe("WorkflowEditorComponent", () => {
       expect(jointGraphWrapper.getCurrentHighlightedOperatorIDs()).toContain(mockScanPredicate.operatorID);
       expect(jointGraphWrapper.getCurrentHighlightedOperatorIDs()).toContain(mockResultPredicate.operatorID);
     });
+
+    //undo
+    it("should undo action when press command + Z or control + Z", () => {
+      spyOn(workflowVersionService, "getDisplayParticularVersionStream").and.returnValue(of(false));
+      spyOn(workflowCollabService, "isLockGranted").and.returnValue(true);
+      spyOn(undoRedoService, "canUndo").and.returnValue(true);
+      let undoSpy = spyOn(undoRedoService, "undoAction");
+      fixture.detectChanges();
+      const event1 = new KeyboardEvent("keydown", { key: "Z", metaKey: true, shiftKey: false });
+      (document.activeElement as HTMLElement)?.blur();
+      document.dispatchEvent(event1);
+      fixture.detectChanges();
+      setTimeout(() => {
+        expect(undoSpy).toHaveBeenCalled();
+      }, 0);
+
+      const event2 = new KeyboardEvent("keydown", { key: "Z", ctrlKey: true, shiftKey: false });
+      (document.activeElement as HTMLElement)?.blur();
+      document.dispatchEvent(event2);
+      fixture.detectChanges();
+      setTimeout(() => {
+        expect(undoSpy).toHaveBeenCalled();
+      }, 0);
+    });
+
+    //redo
+    it("should redo action when press command + Z or control + Z", () => {
+      spyOn(workflowVersionService, "getDisplayParticularVersionStream").and.returnValue(of(false));
+      spyOn(workflowCollabService, "isLockGranted").and.returnValue(true);
+      spyOn(undoRedoService, "canRedo").and.returnValue(true);
+      let redoSpy = spyOn(undoRedoService, "redoAction");
+      fixture.detectChanges();
+      const event1 = new KeyboardEvent("keydown", { key: "y", metaKey: true, shiftKey: false });
```

Comment:
Try to name in a more informative way instead of `event1`.

Reply:
OK, I will make changes after zuozhi reviewing my code.